File: app/src/main/java/com/djfos/im/ui/HomeFragment.java
package com.djfos.im.ui;


import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;
import android.os.Environment;
import android.provider.MediaStore;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;

import com.djfos.im.R;
import com.djfos.im.databinding.FragmentHomeBinding;
import com.djfos.im.viewModel.SharedViewModel;

import java.io.File;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;
import java.util.Objects;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.core.content.ContextCompat;
import androidx.core.content.FileProvider;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProviders;
import androidx.navigation.Navigation;

import static android.app.Activity.RESULT_CANCELED;
import static android.app.Activity.RESULT_OK;


public class HomeFragment extends Fragment {
    private static final String TAG = "HomeFragment";
    private static final int REQUEST_IMAGE_CAPTURE = 1;
    private static final int REQUEST_IMAGE_PICK = 2;
    private static final int PERMISSIONS_WRITE_EXTERNAL_STORAGE = 3;
    private SharedViewModel model;

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        FragmentHomeBinding binding = FragmentHomeBinding.inflate(inflater, container, false);
        model = ViewModelProviders.of(requireActivity()).get(SharedViewModel.class);

        binding.getRoot().<Button>findViewById(R.id.button_choose_image).setOnClickListener((view) -> takePhoto());

        binding.getRoot().<Button>findViewById(R.id.buttonPick).setOnClickListener((view) -> selectPhoto());

        return binding.getRoot();
    }


    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        Log.d(TAG, "onActivityResult() called with: requestCode = [" + requestCode + "], resultCode = [" + resultCode + "], data = [" + data + "]");
        switch (requestCode) {
            case REQUEST_IMAGE_CAPTURE:
                if (resultCode == RESULT_OK) {
                    Uri uri = Uri.fromFile(model.image);
                    Log.d(TAG, "onActivityResult: uri " + uri);
                    Objects.requireNonNull(model.config.getValue()).setUri(uri);
                    model.image = null;
                    Navigation.findNavController(requireView()).navigate(R.id.action_homeFragment_to_adjustPage);
                }
                if (resultCode == RESULT_CANCELED) {
                    model.image = null;
                }
                break;
            case REQUEST_IMAGE_PICK:
                if (resultCode == RESULT_OK) {
                    Uri uri = (data.getData());
                    Log.d(TAG, "onActivityResult: uri " + uri);
                    Objects.requireNonNull(model.config.getValue()).setUri(uri);
                    model.image = null;

                    Navigation.findNavController(requireView()).navigate(R.id.action_homeFragment_to_adjustPage);
                    Log.d(TAG, "onActivityResult: nav ed");
                }
                if (resultCode == RESULT_CANCELED) {
                    model.image = null;
                }
                break;
        }
    }

    private void takePhoto() {
        if (!hasPermission()) {
            requestPermission();
        }

        if (hasPermission()) {
            Intent intent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
            if (intent.resolveActivity(requireActivity().getPackageManager()) != null) {
                File photoFile = null;
                try {
                    photoFile = createImageFile();
                } catch (IOException ex) {
                    // Error occurred while creating the File
                    Log.e(TAG, "capturePhoto: ", ex);
                }
                // Continue only if the File was successfully created
                if (photoFile != null) {
                    Uri photoURI = FileProvider.getUriForFile(requireActivity(), "com.djfos.fileprovider", photoFile);
                    intent.putExtra(MediaStore.EXTRA_OUTPUT, photoURI);
                    model.image = photoFile;
                    startActivityForResult(intent, REQUEST_IMAGE_CAPTURE);
                }
            }
        }
    }

    private File getPhotoPath() {
        File path = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES);
        File appImagePath = new File(path, "im");
        if (!appImagePath.exists()) {
            boolean mkdirs = appImagePath.mkdirs();
        }
        return appImagePath;
    }

    private File createImageFile() throws IOException {
        String timeStamp = new SimpleDateFormat("yyyyMMdd", Locale.CHINESE).format(new Date());
        File appImagePath = getPhotoPath();
        return File.createTempFile(timeStamp, ".jpg", appImagePath);
    }

    private void selectPhoto() {
        Intent intent = new Intent(Intent.ACTION_GET_CONTENT);
        intent.setType("image/*");
        if (intent.resolveActivity(requireActivity().getPackageManager()) != null) {
            startActivityForResult(intent, REQUEST_IMAGE_PICK);
        }
    }

    private boolean hasPermission() {
        return ContextCompat.checkSelfPermission(requireActivity(), Manifest.permission.WRITE_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED;
    }

    private void doPermissionRequest() {
        this.requestPermissions(new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE}, PERMISSIONS_WRITE_EXTERNAL_STORAGE);
    }

    private void requestPermission() {
        if (hasPermission()) {
            return;
        }

        if (shouldShowRequestPermissionRationale(Manifest.permission.WRITE_EXTERNAL_STORAGE)) {
            new AlertDialog.Builder(requireActivity())
                    .setTitle("permission required!")
                    .setMessage("need storage permission to save the photo you take.")
                    .setPositiveButton("OK", (dialog, which) -> doPermissionRequest())
                    .setNegativeButton("Cancel", null)
                    .create()
                    .show();
        } else {
            doPermissionRequest();
        }
    }
}

File: app/src/main/java/com/djfos/im/ui/HomeFragment.kt
package com.djfos.im.ui


import android.app.Activity.RESULT_OK
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.provider.MediaStore
import android.util.Log
import android.view.*
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.FileProvider
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.navigation.fragment.findNavController
import androidx.recyclerview.selection.SelectionPredicates
import androidx.recyclerview.selection.SelectionTracker
import androidx.recyclerview.selection.StorageStrategy
import androidx.recyclerview.widget.LinearLayoutManager
import com.djfos.im.R
import com.djfos.im.adapter.DraftAdapter
import com.djfos.im.adapter.DraftLookup
import com.djfos.im.adapter.RecyclerViewIdKeyProvider
import com.djfos.im.databinding.FragmentHomeBinding
import com.djfos.im.util.*
import com.djfos.im.viewModel.HomePageViewModel
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import org.koin.androidx.viewmodel.ext.android.viewModel
import java.io.File
import java.io.IOException


class HomeFragment : Fragment() {
    private lateinit var mTracker: SelectionTracker<Long>
    private var image: File? = null
    private var actionMode: ActionMode? = null
    private val viewModel by viewModel<HomePageViewModel>()
    private lateinit var draftAdapter: DraftAdapter

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val binding = FragmentHomeBinding.inflate(inflater, container, false)

        // set menu
        (requireActivity() as AppCompatActivity).apply {
            setSupportActionBar(binding.toolbar)
            supportActionBar!!.setDisplayShowTitleEnabled(false)
        }

        draftAdapter = DraftAdapter()

        viewModel.allDrafts.observe({ lifecycle }) { drafts -> draftAdapter.submitList(drafts) }

        binding.draftList.apply {
            setHasFixedSize(true)
            layoutManager = LinearLayoutManager(requireContext())
            adapter = draftAdapter

            // add multi select support
            mTracker = SelectionTracker.Builder<Long>(
                    "draft-list-selection",
                    this,
                    RecyclerViewIdKeyProvider(this),
                    DraftLookup(this),
                    StorageStrategy.createLongStorage()
            ).withSelectionPredicate(
                    SelectionPredicates.createSelectAnything()
            ).build()
        }

        // report the change to viewModel
        mTracker.addObserver(selectionObserver)


        // synchronize multiSelection state between SelectionTracker and actionMode
        viewModel.multiSelection.observe(this, Observer { hasSelection ->
            if (hasSelection) {
                if (actionMode == null) {
                    actionMode = requireActivity().startActionMode(actionModeCallback)
                    Log.d(TAG, "multiSelection start")
                }
            } else {
                actionMode?.apply {
                    finish()
                    mTracker.clearSelection()
                    actionMode = null
                    Log.d(TAG, "multiSelection finish")
                }
            }
        })

        draftAdapter.tracker = mTracker

        setHasOptionsMenu(true)
        return binding.root
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        if (savedInstanceState != null) {
            mTracker.onRestoreInstanceState(savedInstanceState)
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        mTracker.onSaveInstanceState(outState)
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        Log.d(TAG, "onCreateOptionsMenu: ")
        inflater.inflate(R.menu.home_page_menu, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.groupId) {
            R.id.group_new_draft -> when (item.itemId) {
                R.id.action_take_photo -> {
                    takePhoto()
                    true
                }
                R.id.action_pick_from_gallery -> {
                    selectPhoto()
                    true
                }
                else -> super.onOptionsItemSelected(item)
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        when (requestCode) {
            REQUEST_IMAGE_CAPTURE -> if (resultCode == RESULT_OK) {
                val uri = Uri.fromFile(image)
                Log.d(TAG, "onActivityResult: uri $uri")
                image = null
                toAdjustPage(uri)
            }
            REQUEST_IMAGE_PICK -> if (resultCode == RESULT_OK) {
                if (data == null) return
                val uri = data.data ?: return
                requireActivity().contentResolver.takePersistableUriPermission(uri, Intent.FLAG_GRANT_READ_URI_PERMISSION)
                Log.d(TAG, "onActivityResult: uri $uri")
                toAdjustPage(uri)
            }
        }
    }

    private fun toAdjustPage(uri: Uri) {
        GlobalScope.launch {
            val id = viewModel.createDraft(uri.toString())
            findNavController().navigate(HomeFragmentDirections.actionHomeFragmentToAdjustPage(id))
        }
    }


    private fun takePhoto() {
        requestPermission(requireActivity())
        if (!hasPermission(requireActivity()))
            return

        val intent = Intent(MediaStore.ACTION_IMAGE_CAPTURE)
        if (intent.resolveActivity(requireActivity().packageManager) != null) {
            try {
                val photoFile = createImageFile(photoPath)
                val photoURI = FileProvider.getUriForFile(requireActivity(), FILE_PROVIDER, photoFile)
                intent.putExtra(MediaStore.EXTRA_OUTPUT, photoURI)
                image = photoFile
                startActivityForResult(intent, REQUEST_IMAGE_CAPTURE)
            } catch (ex: IOException) {
                Log.e(TAG, "capturePhoto: ", ex)
            }
        }
    }

    private fun selectPhoto() {
        requestPermission(requireActivity())
        if (!hasPermission(requireActivity()))
            return

        val intent = Intent(Intent.ACTION_OPEN_DOCUMENT)

        intent.type = "image/*"
        if (intent.resolveActivity(requireActivity().packageManager) != null) {
            startActivityForResult(intent, REQUEST_IMAGE_PICK)
        }
    }


    private val selectionObserver = object : SelectionTracker.SelectionObserver<Long>() {
        override fun onSelectionRestored() {
            viewModel.multiSelection.value = mTracker.hasSelection()
        }

        override fun onSelectionChanged() {
            viewModel.multiSelection.value = mTracker.hasSelection()
        }
    }

    private val actionModeCallback = object : ActionMode.Callback {

        override fun onCreateActionMode(mode: ActionMode, menu: Menu): Boolean {
            mode.menuInflater.inflate(R.menu.home_page_action_mode, menu)
            return true
        }

        override fun onPrepareActionMode(mode: ActionMode, menu: Menu): Boolean {
            return false // Return false if nothing is done
        }

        override fun onActionItemClicked(mode: ActionMode, item: MenuItem): Boolean {
            return when (item.itemId) {
                R.id.delete -> {
                    viewModel.deleteDrafts(draftAdapter.getSelectedDraft())
                    viewModel.multiSelection.value = false
                    true
                }
                else -> false
            }
        }

        override fun onDestroyActionMode(mode: ActionMode) {

        }
    }


    companion object {
        private const val TAG = "HomeFragment"
        private const val REQUEST_IMAGE_CAPTURE = 1
        private const val REQUEST_IMAGE_PICK = 2
    }
}


